java/src/test/java/org/apache/arrow/gandiva/evaluator/BaseNativeEvaluatorTest.java

```diff
@@ -0,0 +1,243 @@
+    }
+    return buffer;
+  }
+
+  ArrowBuf longBuf(long[] longs) {
+    ArrowBuf buffer = allocator.buffer(longs.length * 8);
+    for (int i = 0; i < longs.length; i++) {
+      buffer.writeLong(longs[i]);
+    }
+    return buffer;
+  }
+
+  ArrowBuf doubleBuf(double[] data) {
+    ArrowBuf buffer = allocator.buffer(data.length * 8);
+    for (int i = 0; i < data.length; i++) {
+      buffer.writeDouble(data[i]);
+    }
+
+    return buffer;
+  }
+
+  ArrowBuf stringToMillis(String[] dates) {
+    ArrowBuf buffer = allocator.buffer(dates.length * 8);
+    for(int i = 0; i < dates.length; i++) {
+      Instant instant = Instant.parse(dates[i]);
+      buffer.writeLong(instant.getMillis());
+    }
+
+    return buffer;
+  }
+
+  void generateData(DataAndVectorGenerator generator, int numRecords, ArrowBuf buffer) {
+    for(int i = 0; i < numRecords; i++) {
+      generator.writeData(buffer);
+    }
+  }
+
+  private long generateDataAndEvaluate(DataAndVectorGenerator generator,
+                                       NativeEvaluator evaluator,
+                                       int numFields, int numExprs,
+                                       int numRows, int maxRowsInBatch,
+                                       int inputFieldSize)
+    throws GandivaException, Exception {
+    int numRemaining = numRows;
+    List<ArrowBuf> inputData = new ArrayList<ArrowBuf>();
+    List<ArrowBuf> dataBufs = new ArrayList<ArrowBuf>();
+    List<ArrowFieldNode> fieldNodes = new ArrayList<ArrowFieldNode>();
+    List<ValueVector> outputVectors = new ArrayList<ValueVector>();
+
+    long start;
+    long finish;
+    long elapsedTime = 0;
+
+    // set the bitmap
+    ArrowBuf validity = arrowBufWithAllValid(maxRowsInBatch);
+    while (numRemaining > 0) {
+      int numRowsInBatch = maxRowsInBatch;
+      if (numRowsInBatch > numRemaining) {
+        numRowsInBatch = numRemaining;
+      }
+
+      // generate data
+      for (int i = 0; i < numFields; i++) {
+        ArrowBuf buf = allocator.buffer(numRowsInBatch * inputFieldSize);
+        generateData(generator, numRowsInBatch, buf);
+
+        fieldNodes.add(new ArrowFieldNode(numRowsInBatch, 0));
+        inputData.add(validity);
+        inputData.add(buf);
+        dataBufs.add(buf);
+      }
+
+      // create record batch
+      ArrowRecordBatch recordBatch = new ArrowRecordBatch(numRowsInBatch, fieldNodes, inputData);
+
+      // set up output vectors
+      // for each expression, generate the output vector
+      for (int i = 0; i < numExprs; i++) {
+        ValueVector valueVector = generator.generateOutputVector(numRowsInBatch);
+        outputVectors.add(valueVector);
+      }
+
+      start = System.nanoTime();
+      evaluator.evaluate(recordBatch, outputVectors);
+      finish = System.nanoTime();
+
+      elapsedTime += (finish - start);
+      // fix numRemaining
+      numRemaining -= numRowsInBatch;
+
+      for(ArrowBuf buf : dataBufs) {
+        buf.release();
```

Comment:
why release databufs separately ?

Reply:
I think we are missing to release the validity buffer here. dataBufs array only contains the buffers allocated inside the loop above while generating data. Validity buffer is not a part of this list and I don't see a release call on that buffer

Reply:
maybe, you need to explicitly close the allocator at the end of the test to see the errors ?